### DISCUSSION TITLE
How to define a wildcard path in a way that lets me access the wildcard part as (input) variable

### DISCUSSION THREAD

**Discussion by BerwoutK** [danielgtaylor/huma#432] [Q&A] [answered]
Hello all, sorry if this is not the right place to ask this but we recently swapped to huma, love it so far however we have a use-case that I can not find how to do.
We have the API endpoint of `/route/{wildcard}` and need it to respond to stuff like wildcard: `foo/bar` if I make it `/route/*` it will listen to the right urls/paths but how do I access anything passed after `/route/`?

huma.Context seems to have the value but that is not accessable in a `huma.Get`

**Comment by x-user** [comment]
Are you talking about something like this?
```go
huma.Get(
	api, "/test/{id}",
	func(
		ctx context.Context,
		input *struct {
			ID string `path:"id"`
		},
	) (*struct{}, error) {
		slog.LogAttrs(
			context.Background(),
			slog.LevelDebug,
			"test route",
			slog.String("id", input.ID),
		)
		return nil, nil
	}
)
```

**Reply by BerwoutK** [reply]
Yes, but the id would be `foo/bar` for example

**Reply by x-user** [reply]
Don't understand you. `id` is just a name for path parameter to get value from underlying router. Nothing more nothing less. Even if router support some pattern matching for it, this has nothing to do with huma. Huma only does basic transformation of path string if router using different syntax for path params than pass it to router. For example for echo `/test/{id}` will become `/test/:id`.

**Reply by BerwoutK** [reply]
Hmm, maybe I can clear it up like this I have the following code:
```go
huma.Get(*API, "/api/*", func(ctx context.Context, input *struct {
	}) (*struct{}, error) {
		// I want to access everything after /api/
		// e.g. if we get /api/fruits/banana I want to get /fruits/banana
		return nil, nil
})
```
	
Does this clear it up? Or should I seek to solve this in the router?

**Reply by x-user** [reply, marked as answer]
This will work if fiber
```go
app := fiber.New()
api := humafiber.New(app, huma.DefaultConfig("API", "0.0.1"))

huma.Get(
	api,
		"/test/+",
	func(
		ctx context.Context,
		input *struct {
			Sub string `path:"+"`
		},
	) (*struct{}, error) {
		slog.LogAttrs(
			context.Background(),
			slog.LevelInfo,
			"test route",
			slog.String("sub", input.Sub),
		)
		return nil, nil
	},
)
```

**Reply by BerwoutK** [reply]
Goated, setting it to `*` and using `path: "*"` did it, thank you!

**Reply by x-user** [reply]
Chi example also working:
```go
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/danielgtaylor/huma/v2"
	"github.com/danielgtaylor/huma/v2/adapters/humachi"
	"github.com/go-chi/chi/v5"
	"github.com/go-chi/chi/v5/middleware"
	"github.com/go-chi/httplog/v2"
)


var (
	port = flag.Uint("port", 8080, "")

	logger = httplog.NewLogger("testapp", httplog.Options{
		JSON:             true,
		LogLevel:         slog.LevelDebug,
		Concise:          true,
		RequestHeaders:   true,
		MessageFieldName: "message",
	})
)

func main() {
	flag.Parse()
	slog.SetDefault(logger.Logger)

	mux := chi.NewMux()

	mux.Use(middleware.RealIP)
	mux.Use(middleware.RequestID)
	mux.Use(httplog.RequestLogger(logger))
	mux.Use(middleware.Recoverer)

	mux.NotFound(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/problem+json")
		w.WriteHeader(http.StatusNotFound)
		fmt.Fprintf(w, `{"title": "Not Found","status": 404, "detail": "Endpoint Not Found"}`)
	})

	api := humachi.New(mux, huma.DefaultConfig("API", "0.0.1"))

	huma.Get(
		api,
		 "/test/*",
		func(
			ctx context.Context,
			input *struct {
				Sub string `path:"*"`
			},
		) (*struct{}, error) {
			slog.LogAttrs(
				context.Background(),
				slog.LevelInfo,
				"test route",
				slog.String("sub", input.Sub),
			)
			return nil, nil
		},
	)

	srv := http.Server{
		Addr:     fmt.Sprintf(":%d", *port),
		ErrorLog: slog.NewLogLogger(logger.Handler(), slog.LevelError),
		Handler: mux,
	}

	if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
		slog.LogAttrs(
			context.Background(),
			slog.LevelError,
			"server error",
			slog.Any("error", err),
		)
	}
}
```

**Reply by mesiriak** [reply]
Doesn't it a things that would be great to put in documentation? It seems not obvious when you start working with framework imo

**Reply by kgantsov** [reply]
It work fine but the UI sends `*` instead of the path when you click on `Send API request` button
<img width="1919" height="1124" alt="Screenshot 2025-12-08 at 13 06 41" src="https://github.com/user-attachments/assets/be74bc73-f940-4240-a7a7-d3d090b4a039" />

**Reply by phrozen** [reply]
I'm having the same issue, but I am using `http.ServeMux` for now, if any other router will help I wouldn't mind switching. Here is my scenario and how it "partially" works:

```go
// Read note
huma.Register(api, huma.Operation{
	OperationID: "read-note",
	Method:      "GET",
	Path:        "/notes/{path}",
	Summary:     "Read a note",
	Description: "Retrieve a note by its path",
	Tags:        []string{"Notes"},
}, s.ReadHandler)
```
This is the naive version, and it only works for Curl or Swagger UI if you URL encode the path like `my%2Fnote%2fpath`. Sending `my/note/path` will return a 404.

Now passing a variadic argument to the router works as expected, capturing the rest of the arguments as input:
```go
// Read note
huma.Register(api, huma.Operation{
	OperationID: "read-note",
	Method:      "GET",
	Path:        "/notes/{path...}",
	Summary:     "Read a note",
	Description: "Retrieve a note by its path",
	Tags:        []string{"Notes"},
}, s.ReadHandler)
```
Now I can curl `/notes/my/note/path` correctly, input will just be `my/note/path`, but now the Swagger UI passes `{path...}` as a string regardless of the input.

I will have to try Chi or other routers to see the differences. Have anyone solved it for both?

**Comment by phrozen** [comment]
It shouldn't be marked answered entirely, although the backend examples work with multiple routers (ServeMux `{var...}`, Fiber `+`, Chi `*`, the ones I tested), in ALL cases the spec is broken, so Swagger UI, Restish and `openapi-typescript-codegen`for example are useless. The whole point of Huma is to have a valid spec, the fix is to have Huma identify those paths and produce a valid spec (which is just a variable name regardless of the router). 

Example:

- ServeMux `/api/{path...}` -> `/api/{path}`
- Chi `/api/*`-> `/api/{sub}`
- Fiber `/api/+`-> `/api/{sub}`

Or whatever the name it doesn't matter as long as the spec does not break, because the spec does not support wildcards, so it has to be interpreted as a path parameter at the end, whatever that parameter is, but no `{path...}`, `+` or `*` as a variable name.

Here is an example of how Connexion framework (Python) solves for this with `format: path`:

```yaml
paths:
  /pages/{path}:
    get:
      summary: Retrieve a page by its path
      operationId: api.get_page
      parameters:
        - name: "path"
          in: path
          description: "Remainder of path, including slashes."
          required: true
          schema:
            type: string
            format: path
      responses:
        '200':
          description: OK
```

And it doesn't have to be solved per adapter, because the path to the endpoint is passed to Huma at registration, Huma can either respect the name in the case of ServeMux `{path...}`or use a placeholder name (`sub` for example) for wilcard paths like Chi's `*`or Fiber's `+` and add the `format` attribute to the schema.  Because as of today, even if wildcards work in Go backends, Huma is broken, because its whole purpose is to produce a valid spec, which it isn't doing.
